### PR TITLE
Deterministic filenames for `write_stan_file`

### PR DESCRIPTION
**Pull request by martinmodrak**
#### Submission Checklist

- [x] Run unit tests
- [x] Declare copyright holder and agree to license (see below)

#### Summary

Makes `write_stan_file` choose filenames deterministically based on the code, so that models do not get unnecessarily recompiled  when calling `write_stan_file` multiple times with the same code, as discussed in #483.

To make this work I needed a hashing algorithm. I chose `rlang::hash`, because we already implicitly depend on `rlang` via `posterior`. The dependency for `rlang` has now been made explicit. I think that's not a big burden, but if needed, the dependency can probably be removed and we can have hashing function implemented locally (it doesn't seem that base R has a hashing function exposed).

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting 
(this will be you or your assignee, such as a university or company): 

Institute of Microbiology of the Czech Academy of Sciences

By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the following licenses:

- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

**Comment by martinmodrak** [comment]
OK, probably should have tagged @mitzimorris earlier, looking at the discussion on @mike-lawrence 's proposals for caching... Anyway, no worries if this gets not included, it was just a bit over 1 hour of implementation, so feel free to push back :-)

**Comment by rok-cesnovar** [comment]
Thanks Marting!

As previously stated, I really like this idea and fully support it. 

I dont think cmdstanpy even has the `write_stan_file` convenience function. We added it because it something that is useful in vignettes and packages that use cmdstanr.

Will review later today or tomorrow the latest.

I think making rlang dependency explicit is not a big deal and certainly prefer using it over implementing our own function.

File: R/example.R [hunk 1]
```diff
@@ -99,7 +106,11 @@ print_example_program <-
 #'   If omitted, a [temporary directory][base::tempdir] is used by default.
 #' @param basename If `dir` is specified, an optional string providing the
 #'   basename for the file created. If not specified a file name is generated
-#'   via [base::tempfile()].
+#'   via [rlang::hash()] of the code.
+#' @param force_overwrite if set to \code{TRUE} the file will always be
+#'   overwritten and thus the resulting model will always be recompiled
+#' @param hash_salt will be added to model code prior hashing to determine
+#'   filename if the \code{basename} param is not set.
```
Comment (jgabry) [inline R/example.R:113]:
```suggestion
#' @param force_overwrite If set to `TRUE` the file will always be
#'   overwritten and thus the resulting model will always be recompiled.
#' @param hash_salt A string that will be added to model code prior to hashing to determine
#'   the file name if the `basename` param is not set.
```

File: R/example.R [hunk 2]
```diff
@@ -128,19 +139,37 @@ print_example_program <-
...
     }
     file <- file.path(dir, basename)
   } else {
-    file <- tempfile(fileext = ".stan", tmpdir = dir)
+    hash <- rlang::hash(paste0(hash_salt, collapsed_code))
+    file <- file.path(dir, paste0("model_", hash, ".stan"))
+  }
+  overwrite <- TRUE
+  # Do not overwrite file if it has the correct contents (to avoid recompilation)
+  if(!force_overwrite && file.exists(file)) {
```
Comment (jgabry) [inline R/example.R:161]:
```suggestion
  if (!force_overwrite && file.exists(file)) {
```

**Review by jgabry — COMMENTED** [review]
Thanks @martinmodrak! I just made a few cosmetic suggestions to match the doc and code style we've been using.

**Comment by rok-cesnovar** [comment]
Is this is good to merge once recent master is merged in @jgabry?

**Comment by jgabry** [comment]
Yeah I think so. I can take care of it now.

**Comment by jgabry** [comment]
Ok I fixed the merge conflicts (just doc) and added an item in NEWS. Will merge now.

**Pull request merged**